custom_types.go

```diff
@@ -0,0 +1,42 @@
+package jsonapi
+
+import "reflect"
+
+type MarshallingFunc func(interface{}) (string, error)
```

Comment:
Follow pattern of `net/http`'s `http.Handle` and `http.HandleFunc`, and don't rely on input/output being string e.g.

```go
type TypeMarshaller interface {
    Marshal(interface{}) (interface{}, error)
}

type TypeUnmarshaller interface {
    UnMarshal(interface{}) (interface{}, error)
}

type TypeMarshalFunc func(interface{}) (interface{}, error)

func (tmf TypeMarshalFunc) func(thing interface{}) (interface{}, error) {
    return tmf(thing)
}

type TypeUnmarshalFunc func(interface{}) (interface{}, error)

func (tuf TypeUnmarshalFunc) func(thing interface{}) (interface{}, error) {
    return tuf(thing)
}
```

Then you can have `RegisterCustomType(in, out reflect.Type, marshaller TypeMarshaller, unmarshaller TypeUnmarshaller) and `RegisterCustomTypeFunc(in, out reflect.Type, marshallerFunc TypeMarshallerFunc, unmarshallerFunc TypeUnmarshallerFunc)`